### PR TITLE
ARROW-7360: [R] Can't use dplyr filter() with variables defined in parent scope

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

File: r/R/dplyr.R [hunk 1]
```diff
@@ -126,7 +124,18 @@ filter.arrow_dplyr_query <- function(.data, ..., .preserve = FALSE) {
   filters <- lapply(filts, function (f) {
     # This should yield an Expression as long as the filter function(s) are
     # implemented in Arrow.
-    try(eval_tidy(f, dm), silent = TRUE)
+    tryCatch(eval_tidy(f, dm), error = function(e) {
+      # Look for the cases where bad input was given, i.e. this would fail
+      # in regular dplyr anyway
+      msg <- conditionMessage(e)
+      if (grepl("object '.*'.not.found", msg)) {
```
Comment (fsaintjacques) [inline r/R/dplyr.R]: Why do you need this?

**Review by fsaintjacques — APPROVED** [review]
I'm just curious at the error matching via string search.

File: r/R/dplyr.R (hunk 1 above)
Comment (nealrichardson) [reply r/R/dplyr.R]: There are two reasons a filter could have failed at this stage: (1) operations that are supported in R and would work in dplyr on a data.frame in R but that aren't (yet) supported in Arrow; and (2) user error, stuff that is invalid even if the data were already in R. These are (at least two of) the ways that the latter case manifests. Previously, we were intercepting variable references that don't exist in the Dataset and erroring earlier, but to allow the scoping to find the other variable you defined in the global environment, I removed that check, let the expression try to evaluate, and catch the exception.

In Python I'd write this something like:

```python
try:
    eval_tidy(f, dm)
except NotFound:
    # User error, so tell them
    raise
except:
    raise ArrowError("Filter expression %s not supported in Arrow" % str(f))
```

R doesn't (generally) have exception classes, so sniffing the exception message is necessary.

File: r/tests/testthat/test-dataset.R [hunk 2]
```diff
@@ -134,10 +134,15 @@ test_that("filter() with %in%", {
...
+  datetime_to_ns <- function (x) {
+    # TODO: src/expression.cpp should handle timestamp data
+    # TODO: C++ library should handle autocasting
+    as.numeric(x) * 1000000
+  }
   expect_equivalent(
     ds %>%
-      # Replace with expression with function once ARROW-7360 is fixed
-      filter(ts >= (as.integer(as.POSIXct(lubridate::ymd_hms("2015-04-29 03:12:39") + lubridate::days(5))) * 1000000), part == 1) %>%
+      filter(ts >= datetime_to_ns(lubridate::ymd_hms("2015-05-04 03:12:39"))) %>%
```
Comment (nealrichardson) [inline r/tests/testthat/test-dataset.R]: @fsaintjacques FYI ☝️

File: r/R/dplyr.R (hunk 1 above)
Comment (bkietz) [reply r/R/dplyr.R]: In that case, it seems advisable to have a GUID here to make the sniffing easier and collisions less likely

Comment (nealrichardson) [reply r/R/dplyr.R]: I don't follow. These aren't exceptions that my code is raising, so I can't choose how they raise or with what messages.

**Pull request closed without merging**